numalogic/connectors/prometheus.py

```diff
@@ -0,0 +1,237 @@
+
+LOGGER = logging.getLogger(__name__)
+MAX_DATA_POINTS: Final[int] = 11000
+_MAX_RECURSION_DEPTH: Final[int] = 10
+_API_ENDPOINT: Final[str] = "/api/v1/query_range"
+
+
+class PrometheusFetcher(DataFetcher):
+    """
+    Class for fetching data as a dataframe from Prometheus or Thanos.
+
+    Args:
+        prometheus_server: Prometheus/Thanos URL
+        scrape_interval_secs: Prometheus scrape interval in seconds
+    """
+
+    __slots__ = ("_endpoint", "_step_secs")
+
+    def __init__(self, prometheus_server: str, scrape_interval_secs: int = 30):
+        super().__init__(prometheus_server)
+        self._endpoint = f"{self.url}{_API_ENDPOINT}"
+        self._step_secs = scrape_interval_secs
+
+    @staticmethod
+    def build_query(metric: str, filters: dict[str, str]) -> str:
+        """Builds a Prometheus query string from metric name and filters."""
+        query = metric
+        if filters:
+            label_filters = ",".join(f"{k}='{v}'" for k, v in filters.items())
+            return f"{metric}{{{label_filters}}}"
+        return query
+
+    def fetch(
+        self,
+        metric_name: str,
+        start: datetime,
+        end: Optional[datetime] = None,
+        filters: Optional[dict[str, str]] = None,
+        return_labels: Optional[list[str]] = None,
+        aggregate: bool = True,
+    ) -> pd.DataFrame:
+        """
+        Fetches data from Prometheus/Thanos and returns a dataframe.
+
+        Args:
+            metric_name: Prometheus metric name
+            start: Start time
+            end: End time
+            filters: Prometheus label filters
+            return_labels: Prometheus label names as columns to return
+            aggregate: Whether to aggregate the data
+            fill_na_value: Value to fill NaNs with
+
+        Returns
+        -------
+            Dataframe with timestamp and metric values
+
+        Raises
+        ------
+            ValueError: If end time is before start time
+            PrometheusFetcherError: If there is an error while fetching data
+            PrometheusInvalidResponseError: If the response from Prometheus is invalid
+            RecursionError: If the recursive depth exceeds the max depth
+        """
+        query = self.build_query(metric_name, filters)
+
+        LOGGER.info("Constructed Prometheus Query: %s", query)
+
+        end_ts, start_ts = self._init_startend_ts(end, start)
+        results = self.query_range(query, start_ts, end_ts)
+
+        df = pd.json_normalize(results)
+        return_labels = [f"metric.{label}" for label in return_labels or []]
+        if df.empty:
+            LOGGER.warning("Query returned no results")
+            return df
+
+        df = self._consolidate_df(df, metric_name, return_labels)
+
+        if aggregate and return_labels:
+            df = self._agg_df(df, metric_name)
+
+        df.sort_values(by=["timestamp"], inplace=True)
+
+        return df
+
+    def raw_fetch(
```

Comment:
Why do we have fetch and raw_fetch doing almost the same thing Or I am just missing something out here?

Reply:
raw fetch gets the raw query, while fetch takes in parameters.

Reply:
Updated doc strings, sorry for the confusion